### PR TITLE
Preserve sharing settings across integration test runs

### PR DESCRIPTION
**Pull request by phil-davis**
## Description
1) Make generic step definitions for ``When the capabilities are retrieved`` and thus be able to hide the repeated detail about doing a ``GET`` and checking for status ``200``.
2) Move ``Given as an "admin"`` into background since it starts every capabilities scenario.
3) Add code that can check an individual setting rather than a whole set, and make a few tests that use that code (so we know it works)
4) Save the existing capabilities at the start of a scenario.
5) Set the desired starting capabilities relevant to the context (pretty much as in master)
6) Put back the saved capabilities on the way out.
7) Various methods added along the way do reduce duplication of lists of "common" "standard" groups of settings.

After this, I can run any suite/feature/scenario by itself and it sets up the sharing/federation settings that it wants to start with, then puts things back the way they were on the way out. This lets me run "random" bits of integration tests in among other tests and playing in the UI trying to track down bugs, without having to think about what happened to settings.

## Motivation and Context
Be nice to people running all different unit, integration and UI tests in their dev environment.

## How Has This Been Tested?
Try starting by first manually turning off "important" settings (like turn off file sharing completely, turn off group sharing). Run just an individual suite or feature from each suite and make sure they pass. Confirm that the previous settings come back as the test exits.

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
Helpful stuff to make integration tests a bit more robust about their environment, and more friendly to everyday developers.

## Checklist:
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

**Comment by individual-it** [comment]
👍 LGTM, @SergioBertolinSG and @davitol should have a look also as they are more familiar with that code

**Comment by phil-davis** [comment]
Rebased and then added in code to save/restore the new capabilities exposed by https://github.com/owncloud/core/pull/28977

**Comment by phil-davis** [comment]
I added sharing settings to the Federated Sharing suite, a least 1 test needs some guaranteed sharing settings - e.g. "Allow apps to use the ShareAPI" plus "Allow resharing". Jenkins should pass this time.

File: tests/integration/features/bootstrap/AppConfiguration.php [hunk 1]
```diff
@@ -32,6 +35,103 @@ public function serverParameterIsSetTo($parameter, $app, $value) {
...
+		return $this->response->xml()->data->capabilities;
+	}
+
+	/**
+	 * @param string $xml of the capabilities
+	 * @param string $capabilitiesApp the name of the "app" in the capabilities response
+	 * @param string $capabilitiesPath the path to the element
+	 * @return string
+	 */
+	public function getParameterValueFromXml(
```
Comment (SergioBertolinSG) [inline tests/integration/features/bootstrap/AppConfiguration.php]: Parameters fit well in a line like in https://github.com/owncloud/core/pull/28970/files#diff-4398f2dd71d3d7276985e0d62f91d937R122

File: tests/integration/features/bootstrap/AppConfiguration.php [hunk 2]
```diff
@@ -32,6 +35,103 @@ public function serverParameterIsSetTo($parameter, $app, $value) {
...
+
+	/**
+	 * @param string $capabilitiesApp the name of the "app" in the capabilities response
+	 * @param string $capabilitiesPath the path to the element
+	 * @return string
+	 */
+	public function getAppParameter(
+		$capabilitiesApp,
+		$capabilitiesPath
+	) {
```
Comment (SergioBertolinSG) [inline tests/integration/features/bootstrap/AppConfiguration.php]: Parameters fit well in a line like in https://github.com/owncloud/core/pull/28970/files#diff-4398f2dd71d3d7276985e0d62f91d937R122

File: tests/integration/features/bootstrap/AppConfiguration.php [hunk 3]
```diff
@@ -32,6 +35,103 @@ public function serverParameterIsSetTo($parameter, $app, $value) {
...
+	 * @Then the capabilities setting of :capabilitiesApp path :capabilitiesPath is :expectedValue
+	 * @param string $capabilitiesApp the name of the "app" in the capabilities response
+	 * @param string $capabilitiesPath the path to the element
+	 * @param string $expectedValue
+	 */
+	public function theCapabilitiesSettingOfAppParameterIs(
+		$capabilitiesApp,
+		$capabilitiesPath,
+		$expectedValue
+	) {
```
Comment (SergioBertolinSG) [inline tests/integration/features/bootstrap/AppConfiguration.php]: Parameters fit well in a line like in https://github.com/owncloud/core/pull/28970/files#diff-4398f2dd71d3d7276985e0d62f91d937R122

**Comment by phil-davis** [comment]
@SergioBertolinSG I went crazy and ran ``phpcs`` for the files touched here and fixed up all the code format things it told me. It is done now, but some things did seem a bit excessive, like it makes me paste in ``@return void`` on all the functions that are step definitions. That is true, but seems over-kill for functions that are just Behat ones.

**Comment by phil-davis** [comment]
Rebased to make sure all is still good with latest master.

File: tests/integration/features/bootstrap/AppConfiguration.php [hunk 4]
```diff
@@ -32,19 +89,126 @@ public function serverParameterIsSetTo($parameter, $app, $value) {
...
 	/**
+	 * @Then the capabilities setting of :capabilitiesApp path :capabilitiesPath is :expectedValue
+	 * @param string $capabilitiesApp the "app" name in the capabilities response
+	 * @param string $capabilitiesPath the path to the element
+	 * @param string $expectedValue
+	 * @return void
+	 */
+	public function theCapabilitiesSettingOfAppParameterIs(
+		$capabilitiesApp, $capabilitiesPath, $expectedValue
+	) {
```
Comment (SergioBertolinSG) [inline tests/integration/features/bootstrap/AppConfiguration.php]: This fits in one line.

File: tests/integration/features/bootstrap/AppConfiguration.php [hunk 5]
```diff
@@ -32,19 +89,126 @@ public function serverParameterIsSetTo($parameter, $app, $value) {
...
+
+	/**
+	 * @param string $xml of the capabilities
+	 * @param string $capabilitiesApp the "app" name in the capabilities response
+	 * @param string $capabilitiesPath the path to the element
+	 * @return string
+	 */
+	public function getParameterValueFromXml(
+		$xml, $capabilitiesApp, $capabilitiesPath
+	) {
```
Comment (SergioBertolinSG) [inline tests/integration/features/bootstrap/AppConfiguration.php]: Fits in one line.

File: tests/integration/features/bootstrap/CapabilitiesContext.php [hunk 6]
```diff
@@ -15,37 +37,99 @@ class CapabilitiesContext implements Context, SnippetAcceptingContext {
 	/**
 	 * @Then /^fields of capabilities match with$/
 	 * @param \Behat\Gherkin\Node\TableNode|null $formData
+	 * @return void
 	 */
-	public function checkCapabilitiesResponse(\Behat\Gherkin\Node\TableNode $formData) {
-		$capabilitiesXML = $this->response->xml()->data->capabilities;
+	public function checkCapabilitiesResponse(
+		\Behat\Gherkin\Node\TableNode $formData
+	) {
```
Comment (SergioBertolinSG) [inline tests/integration/features/bootstrap/CapabilitiesContext.php]: Fits in one line.

**Comment by phil-davis** [comment]
Format changes done and rebased to current master. Let's see if Jenkins and Travis are happy.

File: tests/integration/features/bootstrap/CapabilitiesContext.php [hunk 7]
```diff
@@ -15,37 +37,97 @@ class CapabilitiesContext implements Context, SnippetAcceptingContext {
...
+	 * @return void
+	 */
+	protected function setupAppConfigs() {
+		// Remember the current capabilities
+		$this->getCapabilitiesCheckResponse();
+		$this->savedCapabilitiesXml = $this->getCapabilitiesXml();
+		// Set the required starting values for testing
+		$this->setupCommonSharingConfigs();
+		$this->setupCommonFederationConfigs();
+		$this->modifyServerConfig(
```
Comment (SergioBertolinSG) [inline tests/integration/features/bootstrap/CapabilitiesContext.php]: Fits in one line.

File: tests/integration/features/bootstrap/CapabilitiesContext.php [hunk 8]
```diff
@@ -15,37 +37,97 @@ class CapabilitiesContext implements Context, SnippetAcceptingContext {
...
+			'core', 'shareapi_allow_resharing', 'yes'
+		);
+		$this->modifyServerConfig(
```
Comment (SergioBertolinSG) [inline tests/integration/features/bootstrap/CapabilitiesContext.php]: Fits in one line.

File: tests/integration/features/bootstrap/CapabilitiesContext.php [hunk 9]
```diff
@@ -15,37 +37,97 @@ class CapabilitiesContext implements Context, SnippetAcceptingContext {
...
+			'core', 'shareapi_enforce_links_password', 'no'
+		);
+		$this->modifyServerConfig(
```
Comment (SergioBertolinSG) [inline tests/integration/features/bootstrap/CapabilitiesContext.php]: Fits in one line.

File: tests/integration/features/bootstrap/CapabilitiesContext.php [hunk 10]
```diff
@@ -15,37 +37,97 @@ class CapabilitiesContext implements Context, SnippetAcceptingContext {
...
+			'core', 'shareapi_allow_public_notification', 'no'
+		);
+		$this->modifyServerConfig(
```
Comment (SergioBertolinSG) [inline tests/integration/features/bootstrap/CapabilitiesContext.php]: Fits in one line.

File: tests/integration/features/bootstrap/CapabilitiesContext.php [hunk 11]
```diff
@@ -15,37 +37,97 @@ class CapabilitiesContext implements Context, SnippetAcceptingContext {
...
+			'core', 'shareapi_allow_social_share', 'yes'
+		);
+		$this->modifyServerConfig(
```
Comment (SergioBertolinSG) [inline tests/integration/features/bootstrap/CapabilitiesContext.php]: Fits in one line.

File: tests/integration/features/bootstrap/CapabilitiesContext.php [hunk 12]
```diff
@@ -15,37 +37,97 @@ class CapabilitiesContext implements Context, SnippetAcceptingContext {
...
+			'core', 'shareapi_default_expire_date', 'no'
+		);
+		$this->modifyServerConfig(
```
Comment (SergioBertolinSG) [inline tests/integration/features/bootstrap/CapabilitiesContext.php]: Fits in one line.

File: tests/integration/features/bootstrap/AppConfiguration.php [hunk 13]
```diff
@@ -27,19 +80,122 @@ public function serverParameterIsSetTo($parameter, $app, $value) {
...
+	}
+
+	/**
+	 * @When the capabilities are retrieved
+	 * @return void
+	 */
+	public function getCapabilitiesCheckResponse() {
+		$this->sendingTo('GET', '/cloud/capabilities');
+
+		PHPUnit_Framework_Assert::assertEquals(
```
Comment (SergioBertolinSG) [inline tests/integration/features/bootstrap/AppConfiguration.php]: Fits in one line.

**Review by SergioBertolinSG — APPROVED** [review]

**Comment by phil-davis** [comment]
Backport stable10 #29176

**Comment by phil-davis** [comment]
Note: the elapsed time on Jenkins for the integration tests went up over 1 hour when setting-resetting all the sharing config at every scenario. (compared to normally around 50 minutes) The last commit just added checks and only actually makes the config setting change if the config is not already set to the desired value, and only resets it if it was changed at the start of the scenario. This avoids a lot of unnecessary API calls to set these settings (which take time) vs doing some simple computational checks.

**Pull request merged**